book_source/06_reference/09_rabbitmq.Rmd

```diff
@@ -0,0 +1,168 @@
+Note the use of [`ARG`](https://docs.docker.com/engine/reference/builder/#arg) here to specify a default version model version of 136 while allowing this to be configurable (via `--build-arg MODEL_VERSION=X`) at build time:
+
+```dockerfile
+ARG MODEL_VERSION=136
+
+ENV APPLICATION="./job.sh" \
+    MODEL_TYPE="SIPNET" \
+    MODEL_VERSION="${MODEL_VERSION}"
+
+ENV RABBITMQ_QUEUE="${MODEL_TYPE}_${MODEL_VERSION}"
+```
+
+**WARNING**: Dockerfile environment variables set via `ENV` are assigned _all at once_; _they do not evaluate successively, left to right_.
+Consider the following block:
+
+```dockerfile
+# Don't do this!
+ENV MODEL_TYPE="SIPNET" \
+    MODEL_VERSION=136 \
+    RABBITMQ_QUEUE=${MODEL_TYPE}_${MODEL_VERSION}   # <- Doesn't know about MODEL_TYPE or MODEL_VERSION!
+```
+
+In this block, the expansion for setting `RABBITMQ_QUEUE` _is not aware_ of the current values of `MODEL_TYPE` or `MODEL_VERSION`, and will therefore be set incorrectly to just `_` (unless they have been set previously, in which case it will be aware only of their earlier values).
+As such, **variables depending on other variables must be set in a separate, subsequent `ENV` statement than the variables they depend on**.
+
+## Case study: PEcAn web interface {#rabbitmq-case-study}
+
+The following describes in general terms what happens during a typical run of the PEcAn web interface with RabbitMQ.
+
+1. The user initializes all containers with `docker-compose up`.
+All the services that interact with RabbitMQ (`executor` and all models) run `receiver.py` in the foreground, waiting for messages to tell them what to do.
+
+2. The user browses to http://localhost:8000/pecan/ and steps through the web interface.
+All the pages up to the `04-runpecan.php` run on the `web` container, and are primarily for setting up the `pecan.xml` file.
+
+3. Once the user starts the PEcAn workflow at `04-runpecan.php`, the underlying PHP code connects to RabbitMQ (based on the URI provided in `config.php`) and posts the following message to the `pecan` queue:
+
+  ```json
+  {"folder": "/workflows/PEcAn_WORKFLOWID", "workflowid": "WORKFLOWID"}
+  ```
+  
+4. The `executor` service, which is listening on the `pecan` queue, hears this message and executes its `APPLICATION` (`R CMD BATCH workflow.R`) in the working directory specified in the message's `folder`.
+   The `executor` service then performs the pre-execution steps (e.g. trait meta-analysis, conversions) itself.
+   Then, to actually execute the model, `executor` posts the following message to the target model's queue:
+
+  ```json
+  {"folder": "/workflows/PEcAn_WORKFLOWID/run/RUNID"}
+  ```
+
+5. The target model service, which is listening on its dedicated queue, hears this message and runs its `APPLICATION`, which is `job.sh`, in the directory indicated by the message.
+   If model execution was successful, the model service posts an OK message to the `pecan` queue.
```

Comment:
Right now it waits for the `rabbitmq.out` file in the folder. Once this file appears it assumes the run was completed. This will will contain the status of the actual run.

Future work will move this to a response on RabbitMQ.

Reply:
fe60c88e9